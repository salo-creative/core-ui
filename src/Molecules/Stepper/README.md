# Stepper

This can be used for any stepped content. Especially useful for complex data forms

---

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
import Stepper from '@salo/core-ui/Molecules/Stepper';
```

Implement as follows

```javascript
<Stepper
  activeItem={ step } // Pass in the active item from state
  changeStep={ (step) => setState(step) } // Set new page of stepper to state
>
  { [
    {
      id: 'step_1', // the id used for changing step
      title: 'First step', // title for the stepper nav
      complete: true, // whether the step should render as complete
      disabled: false, // if the step is disabled
      content: ( // React components to render
        <React.Fragment>
          <P> Some lovely step content</P>
        </React.Fragment>
      )
    },
    {
      id: 'step_2',
      title: 'Second step',
      complete: false,
      content: (
        <React.Fragment>
          <P>Some more lovely step content</P>
        </React.Fragment>
      )
    },
    {
      id: 'step_3',
      title: 'Third step',
      disabled: true,
      complete: false,
      content: (
        <React.Fragment>
          <P>Some more lovely step content</P>
        </React.Fragment>
      )
    }
  ] }
</Stepper>
```

**For full prop types and usage see storybook info/knobs**

