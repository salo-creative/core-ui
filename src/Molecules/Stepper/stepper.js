import React from 'react';
import PropTypes from 'prop-types';
import { isEmpty, get, find, hasIn } from 'lodash';
import P from '../../Typography/P';

// COMPONENTS & STYLES
import { Container } from './stepper.styles';
import Navigation from './stepper.navigation';

const Stepper = (props) => {
  const {
    activeItem,
    changeStep,
    children,
    className
  } = props;

  const [visited, setVisited] = React.useState([activeItem]);

  React.useEffect(() => {
    // Set new active id to state
    if (!visited.includes(activeItem)) {
      setVisited([...visited, activeItem]);
    }
  }, [activeItem, visited]);

  if (isEmpty(children)) {
    return (
      <P>No steps defined</P>
    );
  }

  const activeStep = activeItem || get(children, '[0].id');

  const returnContent = () => {
    const body = find(children, { id: activeStep });
    if (body && hasIn(body, 'content')) {
      return body.content;
    }
    return <P>Active step has no content</P>;
  };

  return (
    <Container className={ className }>
      <Navigation
        activeItem={ activeStep } // use passed in active item or first step
        changeStep={ changeStep }
        steps={ children.map(item => ({
          title: item.title,
          id: item.id,
          complete: item.complete,
          disabled: item.disabled,
          visited: visited.includes(item.id)
        })) }
      />
      { returnContent() }
    </Container>
  );
};

Stepper.defaultProps = {
  activeItem: null,
  children: [],
  className: 'stepper'
};

Stepper.propTypes = {
  activeItem: PropTypes.string,
  changeStep: PropTypes.func.isRequired,
  children: PropTypes.arrayOf(PropTypes.shape({
    title: PropTypes.string.isRequired,
    disabled: PropTypes.bool,
    complete: PropTypes.bool,
    id: PropTypes.string.isRequired,
    content: PropTypes.any.isRequired
  })),
  className: PropTypes.string
};

export default Stepper;