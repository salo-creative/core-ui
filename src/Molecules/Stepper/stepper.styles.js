import styled from 'styled-components';
import { transparentize } from 'polished';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
`;

export const NavWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  position: relative;
  width: 100%;
  margin: 0 0rem 3rem;
  &:before {
    content: '';
    height: 1px;
    left: 0;
    top: 50%;
    position: absolute;
    width: 100%;
    z-index: 1;
    transform: translateY(-50%);
    background: ${ ({ theme }) => theme.grey };
  }
`;

export const NavItem = styled.button`
  transition: all 0.3s linear;
  display: flex;
  border: none;
  background: none;
  align-items: center;
  font-size: 1.2rem;
  padding: 0.5rem 1rem;
  background: #fff;
  cursor: pointer;
  position: relative;
  z-index: 2;
  color: ${ ({ theme }) => theme.font };
  &:before {
    transition: all 0.3s linear;
    content: '';
    display: flex;
    width: 2.5rem;
    height: 2.5rem;
    margin-right: 0.5rem;
    border-radius: 50%;
    background: ${ ({ theme }) => theme.grey };
  }
  &[disabled] {
    color: ${ ({ theme }) => transparentize(0.25, theme.grey) };
    cursor: not-allowed;
    &:before {
      background: ${ ({ theme }) => transparentize(0.25, theme.grey) };
    }
  }
  &.active,
  &.complete {
    &:before {
      background: ${ ({ theme }) => theme.primary };
    }
  }
`;

export const StepWrapper = styled.div`
    width: 100%;
`;