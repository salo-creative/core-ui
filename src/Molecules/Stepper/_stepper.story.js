import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, text } from '@storybook/addon-knobs';
import { State, Store } from '@sambego/storybook-state';

// Tests.
import { withTests } from '@storybook/addon-jest';
import results from '../../../.storybook/jest-test-results.json';

// Featured Component.
import { Stepper } from '../../index';
import P from '../../Typography/P';

// README
import README from './README.md';

const stories = storiesOf('Molecules | Stepper', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withTests({ results }));
stories.addParameters({ jest: ['stepper'] });

stories.add(
  'Basic',
  (() => {
    // Store
    const store = new Store({ step: 'step_1' });
    return (
      <State store={ store }>
        { state => (
          <Stepper
            activeItem={ state.step }
            changeStep={ (id) => store.set({ step: id }) }
          >
            { [
              {
                id: 'step_1',
                title: 'First step',
                complete: true,
                content: (
                  <React.Fragment>
                    <P> Some lovely step content</P>
                  </React.Fragment>
                )
              },
              {
                id: 'step_2',
                title: 'First step',
                complete: true,
                content: (
                  <React.Fragment>
                    <P> Some lovely step content</P>
                  </React.Fragment>
                )
              },
              {
                id: 'step_3',
                title: 'Third step',
                complete: false,
                content: (
                  <React.Fragment>
                    <P>Some more lovely step content</P>
                  </React.Fragment>
                )
              },
              {
                id: 'step_4',
                title: 'Fourth step',
                disabled: true,
                complete: false,
                content: (
                  <React.Fragment>
                    <P>Some more lovely step content</P>
                  </React.Fragment>
                )
              }
            ] }
          </Stepper>
        ) }
      </State>
    );
  }),
  {
    info: { propTables: [Stepper], propTablesExclude: [State, P, React.Fragment] },
    notes: README
  }
);