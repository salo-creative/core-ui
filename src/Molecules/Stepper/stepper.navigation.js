import React from 'react';
import PropTypes from 'prop-types';

// COMPONENTS & STYLES
import { NavWrapper, NavItem } from './stepper.styles';

const Navigation = (props) => {
  const {
    activeItem,
    changeStep,
    steps
  } = props;
  return (
    <NavWrapper>
      { steps.map(step => {
        const active = activeItem === step.id;
        return (
          <NavItem
            key={ step.id }
            className={ `${ active ? 'active' : '' } ${ step.complete ? 'complete' : '' }` }
            onClick={ () => changeStep(step.id) }
            disabled={ step.disabled }
          >
            { step.title }
          </NavItem>
        );
      }) }
    </NavWrapper>
  );
};

Navigation.defaultProps = {};

Navigation.propTypes = {
  activeItem: PropTypes.string.isRequired,
  changeStep: PropTypes.func.isRequired,
  steps: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    complete: PropTypes.bool,
    disabled: PropTypes.bool
  })).isRequired
};

export default Navigation;