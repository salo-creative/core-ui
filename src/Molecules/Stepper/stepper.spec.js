import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import Stepper from './index';

test('Renders against snapshot', async () => {
  const { container } = renderWithTheme(
    <Stepper />
  );
  expect(container.firstChild).toMatchSnapshot();
});