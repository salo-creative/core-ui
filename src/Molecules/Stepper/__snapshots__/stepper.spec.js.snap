// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c0 {
  font-weight: 400;
  width: 100%;
  color: #333333;
  text-align: left;
  margin: 0 0 20px;
  padding: 0;
  font-size: 14px;
  line-height: 17px;
}

.c0 em,
.c0 i {
  font-style: italic;
}

.c0 b,
.c0 strong {
  font-weight: bold;
}

.c0 a {
  font-size: 1em;
  color: currentColor;
  -webkit-transition: color 0.3s;
  transition: color 0.3s;
  font-weight: 600;
}

.c0 a:hover {
  -webkit-transition: color 0.3s;
  transition: color 0.3s;
  color: #8f8f8f;
}

.c0 pre {
  font-family: "Courier 10 Pitch",Courier,monospace;
  font-size: 95%;
  line-height: 140%;
  white-space: pre;
  white-space: -moz-pre-wrap;
  white-space: -o-pre-wrap;
  white-space: pre-wrap;
}

.c0 code {
  font-family: Monaco,Consolas,"Andale Mono","DejaVu Sans Mono",monospace;
  font-size: 95%;
  line-height: 140%;
  white-space: pre;
  white-space: -moz-pre-wrap;
  white-space: -o-pre-wrap;
  white-space: pre-wrap;
  background: #faf8f0;
}

@media screen and (min-width:400px) {
  .c0 {
    font-size: calc(14px + (17 - 14) * (100vw - 400px) / (1200 - 400));
    line-height: calc(17px + (29 - 17) * (100vw - 400px) / (1200 - 400));
  }
}

@media screen and (min-width:1200px) {
  .c0 {
    font-size: 17px;
    line-height: 29px;
  }
}

<p
  class="c0"
  color=""
  font-size=""
>
  No steps defined
</p>
`;
